diff -ur ircd-hybrid/include/s_conf.h new/include/s_conf.h
--- ircd-hybrid/include/s_conf.h	Tue May  6 19:46:46 2003
+++ new/include/s_conf.h	Wed May  7 00:01:04 2003
@@ -244,6 +244,7 @@
   int           throttle_time;
   int           use_egd;
   int		ping_cookie;
+  int		disable_auth;
   int		fallback_to_ip6_int;
 #ifdef HAVE_LIBCRYPTO
   struct EncCapability *default_cipher_preference;
diff -ur ircd-hybrid/src/ircd_lexer.l new/src/ircd_lexer.l
--- ircd-hybrid/src/ircd_lexer.l	Tue May  6 19:46:49 2003
+++ new/src/ircd_lexer.l	Wed May  7 00:01:20 2003
@@ -156,6 +156,7 @@
 deny		{ return DENY; }
 description	{ return DESCRIPTION; }
 die		{ return DIE; }
+disable_auth    { return DISABLE_AUTH; }
 disable_hidden	{ return DISABLE_HIDDEN; }
 disable_local_channels { return DISABLE_LOCAL_CHANNELS; }
 disable_remote_commands	{ return DISABLE_REMOTE_COMMANDS; }
diff -ur ircd-hybrid/src/ircd_parser.y new/src/ircd_parser.y
--- ircd-hybrid/src/ircd_parser.y	Tue May  6 19:46:49 2003
+++ new/src/ircd_parser.y	Wed May  7 00:02:21 2003
@@ -122,6 +122,7 @@
 %token  DENY
 %token  DESCRIPTION
 %token  DIE
+%token  DISABLE_AUTH
 %token  DISABLE_HIDDEN
 %token  DISABLE_LOCAL_CHANNELS
 %token  DISABLE_REMOTE_COMMANDS
@@ -2054,7 +2055,7 @@
                     general_compression_level | general_client_flood |
                     general_throttle_time | general_havent_read_conf |
                     general_dot_in_ip6_addr | general_ping_cookie |
-                    general_fallback_to_ip6_int | 
+                    general_disable_auth | general_fallback_to_ip6_int | 
                     error;
 
 general_failed_oper_notice: FAILED_OPER_NOTICE '=' TYES ';'
@@ -2459,6 +2460,15 @@
   if (ypass == 2)
     ConfigFileEntry.ping_cookie = 0;
 };
+
+general_disable_auth: DISABLE_AUTH '=' TYES ';'
+{
+  ConfigFileEntry.disable_auth = 1;
+} |
+  DISABLE_AUTH '=' TNO ';'
+{
+  ConfigFileEntry.disable_auth = 0;
+} ;
 
 general_throttle_time: THROTTLE_TIME '=' timespec ';'
 {
diff -ur ircd-hybrid/src/s_auth.c new/src/s_auth.c
--- ircd-hybrid/src/s_auth.c	Wed Apr 23 10:40:09 2003
+++ new/src/s_auth.c	Wed May  7 00:02:59 2003
@@ -425,7 +425,9 @@
 		   client->localClient->dns_query, 0) == 0)
     SetDNSPending(auth);
 
-  (void)start_auth_query(auth);
+  if(ConfigFileEntry.disable_auth == 0)
+    (void)start_auth_query(auth);
+
   link_auth_request(auth, &auth_poll_list);
 }
 
