diff -ur ircd-hybrid/doc/example.conf disabled/doc/example.conf
--- ircd-hybrid/doc/example.conf	Sat May 31 00:30:25 2003
+++ disabled/doc/example.conf	Mon Nov 29 19:59:39 2004
@@ -612,6 +612,9 @@
  * in options in config.h.  The general block is read at start time.
  */
 general {
+	/* disable_auth: disable ident checking? */
+	disable_auth = yes;
+
 	/* floodcount: the default value of floodcount that is configurable
 	 * via /quote set floodcount.  This is the amount of lines a user
 	 * may send to any other user/channel in one second.
diff -ur ircd-hybrid/doc/example.conf.quick disabled/doc/example.conf.quick
--- ircd-hybrid/doc/example.conf.quick	Sat May 31 00:30:25 2003
+++ disabled/doc/example.conf.quick	Mon Nov 29 19:59:59 2004
@@ -278,6 +278,7 @@
 };
 
 general {
+        disable_auth = yes;
     default_floodcount = 10;
 	failed_oper_notice = yes;
 	dots_in_ident=2;
diff -ur ircd-hybrid/include/s_conf.h disabled/include/s_conf.h
--- ircd-hybrid/include/s_conf.h	Wed Mar 17 19:57:21 2004
+++ disabled/include/s_conf.h	Mon Nov 29 19:56:30 2004
@@ -241,6 +241,7 @@
   int           throttle_time;
   int           use_egd;
   int		ping_cookie;
+  int		disable_auth
   int           use_help;
 #ifdef IPV6  
   int		fallback_to_ip6_int;
diff -ur ircd-hybrid/src/ircd_lexer.l disabled/src/ircd_lexer.l
--- ircd-hybrid/src/ircd_lexer.l	Sat Oct 25 19:08:22 2003
+++ disabled/src/ircd_lexer.l	Mon Nov 29 19:53:55 2004
@@ -156,6 +156,7 @@
 deny		{ return DENY; }
 description	{ return DESCRIPTION; }
 die		{ return DIE; }
+disable_auth    { return DISABLE_AUTH; }
 disable_hidden	{ return DISABLE_HIDDEN; }
 disable_local_channels { return DISABLE_LOCAL_CHANNELS; }
 disable_remote_commands	{ return DISABLE_REMOTE_COMMANDS; }
diff -ur ircd-hybrid/src/ircd_parser.y disabled/src/ircd_parser.y
--- ircd-hybrid/src/ircd_parser.y	Sat Oct 25 19:08:23 2003
+++ disabled/src/ircd_parser.y	Mon Nov 29 19:55:42 2004
@@ -118,6 +118,7 @@
 %token  DENY
 %token  DESCRIPTION
 %token  DIE
+%token  DISABLE_AUTH
 %token  DISABLE_HIDDEN
 %token  DISABLE_LOCAL_CHANNELS
 %token  DISABLE_REMOTE_COMMANDS
@@ -1989,7 +1990,7 @@
                     general_compression_level | general_client_flood |
                     general_throttle_time | general_havent_read_conf |
                     general_dot_in_ip6_addr | general_ping_cookie |
-                    general_fallback_to_ip6_int | 
+                    general_fallback_to_ip6_int | general_disable_auth
                     error;
 
 general_failed_oper_notice:   FAILED_OPER_NOTICE '=' TYES ';'
@@ -2356,6 +2357,15 @@
     PING_COOKIE '=' TNO ';'
   {
     ConfigFileEntry.ping_cookie = 0;
+  } ;
+
+general_disable_auth: DISABLE_AUTH '=' TYES ';'
+  {
+    ConfigFileEntry.disable_auth = 1;
+  } |
+    DISABLE_AUTH '=' TNO ';'
+  {
+    ConfigFileEntry.disable_auth = 0;
   } ;
 
 general_use_help: USE_HELP '=' TYES ';'
diff -ur ircd-hybrid/src/s_auth.c disabled/src/s_auth.c
--- ircd-hybrid/src/s_auth.c	Wed May 28 22:13:26 2003
+++ disabled/src/s_auth.c	Mon Nov 29 19:55:06 2004
@@ -419,7 +419,9 @@
 		   client->localClient->dns_query, 0) == 0)
     SetDNSPending(auth);
 
-  (void)start_auth_query(auth);
+  if (ConfigFileEntry.disable_auth == 0)
+    start_auth_query(auth);
+
   link_auth_request(auth, &auth_poll_list);
 }
 
