$Id: vprintf_fix.patch,v 1.1 2005/07/02 01:44:07 db Exp $
Index: ircd-hybrid/src/send.c
===================================================================
RCS file: /cvsroot/ircd-hybrid/src/send.c,v
retrieving revision 7.288
diff -u -r7.288 send.c
--- ircd-hybrid/src/send.c	27 Jun 2005 22:13:50 -0000	7.288
+++ ircd-hybrid/src/send.c	2 Jul 2005 01:30:20 -0000
@@ -63,6 +63,7 @@
 send_format(char *lsendbuf, int bufsize, const char *pattern, va_list args)
 {
   int len;
+  va_list args_copy;
 
   /*
    * from rfc1459
@@ -76,7 +77,10 @@
    * continuation message lines.  See section 7 for more details about
    * current implementations.
    */
-  len = vsnprintf(lsendbuf, bufsize - 1, pattern, args);
+  
+  va_copy(args_copy, args);
+  len = vsnprintf(lsendbuf, bufsize - 1, pattern, args_copy);
+  va_end(args_copy);
   if (len > bufsize - 2)
     len = bufsize - 2;  /* required by some versions of vsnprintf */
 
